src/states/node/mod.rs

```diff
@@ -1729,7 +1729,10 @@ impl Node {
         outbox: &mut EventBox,
         try_reconnect: bool,
     ) -> bool {
-        if self.peer_mgr.remove_peer(&pub_id) {
+        // Calling remove twice to remove a potential JoiningNode we have as a Node than purely
+        // demoting it
+        // TODO: Avoid calling this twice.
+        if self.peer_mgr.remove_peer(&pub_id) || self.peer_mgr.remove_peer(&pub_id) {
```

Comment:
This is not ideal (but the TODO suggest you are aware of it). Maybe it would be better to have two functions: `demote_peer` and `remove_peer` ? Or add a param to `remove_peer` which indicates whether to remove or just demote (could be just `bool`, but an `enum` would be better). Another alternative would be to only change the return type of `remove_peer` to indicate whether the peer was removed or only demoted. That would make this `if` condition a bit messier, but at least it would be more explicit about what's going on.

Reply:
Added remove_peer_no_joining_checks.